File: 279-perfect-squares/perfect-squares.py
import math

class Solution:
    def numSquares(self, n: int) -> int:
        # build list of all perfect squares <= n
        perfectSquares = [i * i for i in range(1, int(math.sqrt(n)) + 1)]
        memo = {}

        def DP(num):
            # base case
            if num == 0:
                return 0
            if num < 0:
                return float('inf')

            # check if subproblem computed
            if num in memo:
                return memo[num]

            best = float('inf')
            
            for square in perfectSquares:
                best = min(best, 1 + DP(num - square))

            memo[num] = best
            return best

        return DP(n)

File: 279-perfect-squares/perfect-squares.cpp
class Solution {
public:
    int numSquares(int n) {
        vector<int> squares;
        std::unordered_map<int,int> memo;

        for(int i {1}; i*i <= n; i++)
        {
            squares.push_back(i*i);
        }

        return DP(n,squares,memo);
    }

    int DP(int num, vector<int>& squares, std::unordered_map<int,int>& memo)
    {
        int min_cost = INT_MAX;
        if(num == 0) return 0;

        if(memo.count(num)) return memo[num];

        for(int square : squares)
        {
            if(num - square >= 0)
            {
                min_cost = min(min_cost, 1 + DP(num-square, squares, memo));
            }
        }

        return memo[num] = min_cost;
    }
};